### PR TITLE
CI Rework: More Aggressive Culling Of FPGA Resources, Slack/PR Notifications

### PR DESCRIPTION
**Pull request by abejgonzalez**
This PR attempts to fix/alleviate some CI issues that have caused resource leakages (of both manager and FPGA instances).

Existing issues found:
- The workflow monitor did not start properly (a combination of needing all env. variables, and other small oversights). However, this wasn't apparent since there was no logging/notification of something going wrong.
- FPGA instances were left alive for too long (workflow monitor was failing is one reason for this).

It does the following:

- `workflow_monitor.py` and `setup_workflow_monitor.py`
  - The monitor needs access to all environment variables used in `ci_variables.py`. Thus, instead of setting an `env.sh` file that holds needed env. variables, a new `ci_env` dict is created (see the `ci_variables.py` changes) and set for the corresponding `run()` calls.
  - Now there are 2 verification steps to ensure that the workflow monitor at least started properly (the workflow monitor has it's own check to ensure it is properly running)
  - Since the workflow monitor has access to all env. variables needed, it doesn't have the args anymore.
  - The workflow monitor now sends PR comments when starting + failing
  - The workflow monitor now has a `check_and_terminate_select_instances` that has a 45m timeout on FPGA instances (see `platform_lib.py` for more info).
  - The workflow monitor now has a `try/catch` to check if failures happen (at least it should hopefully send the PR comment if something fails).
- `ci_variables.py` - A new `ci_env` dict is populated with all env. variables used throughout CI. Instead of using python variable names for environment variables (which don't have the same naming convention) use the env. variable name for the key in the dict. All code locations are changed to account for this. This dict is type checked so that all keys are considered present. Future CI type checking will happen in a followup PR.
- `firesim-cleanup.yml` - This workflow now sends a Slack message if something goes wrong. This is tied to the `cull-instances` script now returning an error if it culls something it expected not to cull.
- `cull-old-ci-insts.py` - This now explicitly checks for FPGA instances that are live for over 1hr. Also now exits non-zero if something was terminated unexpectedly (triggers a slack message).
- `{github_}common.py` - Moved GitHub specific items to a new file `github_common.py` to avoid a circular dependency with the `platform_lib` file. A new issues URL and issue function is added for PR comment posts.
- `platform_lib.py`:
  - New `find_select_ci_instances` func. for each platform to find "select" CI instances (i.e. FPGA instances) and return them.
  - New `check_and_terminate_select_instances` func. to find "select" instances and terminate them. If instances are terminated then it sends a PR comment.

There are now multiple checks for FPGA instances (for extra extra extra security):
- Linux test timeout - 30m timeout
- Workflow Monitor - 45m timeout
- Cull CI Instances - 1h timeout

#### Related PRs / Issues

N/A

#### UI / API Impact

N/A

#### Verilog / AGFI Compatibility

N/A

### Contributor Checklist
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [ ] If applicable, did you regenerate and publicly share default AGFIs?

- [x] If applicable, did you apply the `ci:fpga-deploy` label?

- [ ] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
Note: to run CI on PRs from forks, comment `@Mergifyio copy main` and manage the change from the new PR.
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

**Comment by abejgonzalez** [comment]
HI!

**Comment by abejgonzalez** [comment]
HI!

**Comment by abejgonzalez** [comment]
Started up the workflow monitor for CI run {ci_env['GITHUB_RUN_ID']}

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run {ci_env['GITHUB_RUN_ID']}. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
Started up the workflow monitor for CI run 3625480853

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3625480853. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
Started up the workflow monitor for CI run 3625644489

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3625644489. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3626479770

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3626479770. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3626585668

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3626585668. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3626724022

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3626724022. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3626878854

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3626878854. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3627189385

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3627189385. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3627493853

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3627493853. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Comment by abejgonzalez** [comment]
While not working 100%. This PR is now ready for the 1st round of reviews.

Things left TODO:
- Send workflow monitor log as a PR comment when a failure occurs?
- Fix Linux test (Vitis + Normal)

**Comment by t14916** [comment]
This may be adjacent to this PR in particular, but if the aim is to reduce CI costs, you can also try a more aggressive filtering approach right? For example, I think manager/python library changes will run the full gamut of scala tests and chipyard tests, but is that really necessary?

File: .github/scripts/setup-workflow-monitor.py
Comment (t14916) [inline .github/scripts/setup-workflow-monitor.py:12]: This was the only place AZURE_CREDITED_ENV was set

File: .github/scripts/setup-workflow-monitor.py
Comment (t14916) [inline .github/scripts/setup-workflow-monitor.py:28]: since you've removed the env.sh buffoonery I did, you don't need the azure credited env anymore.

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3633039036

**Comment by abejgonzalez** [comment]
> This may be adjacent to this PR in particular, but if the aim is to reduce CI costs, you can also try a more aggressive filtering approach right? For example, I think manager/python library changes will run the full gamut of scala tests and chipyard tests, but is that really necessary?

I agree. However, I'm treating that as future work

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3633720674

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3633857500

**Comment by abejgonzalez** [comment]
```
Workflow monitor started for CI run: 3633857500
```

File: .github/scripts/workflow-monitor.py
Comment (abejgonzalez) [inline .github/scripts/workflow-monitor.py:49]: Thoughts on not doing this (every commit added will send this message polluting the comment stream).

File: .github/scripts/workflow-monitor.py
Comment (t14916) [inline .github/scripts/workflow-monitor.py:76]: Since you're catching exceptions the exceptions here, it might be better to raise an expression with the print message inside of it so that it can be seen from the PR comment, otherwise you're going to have a value error raised and follow the stack trace to see what happened. Something like `raise ValueError(f"Unexpected Workflow State on Completed: {state_concl}")` etc.

File: .github/scripts/cull-old-ci-instances.py
Comment (abejgonzalez) [inline .github/scripts/cull-old-ci-instances.py:57]: Exit w/ non-zero to send a Slack notification

File: .github/scripts/github_common.py
Comment (abejgonzalez) [inline .github/scripts/github_common.py:56]: This is a new function (with a corresponding new `gh_issues_api_url`). The rest of this file is copied.

File: .github/scripts/workflow-monitor.py
Comment (abejgonzalez) [inline .github/scripts/workflow-monitor.py:89]: This is a tradeoff on what we think will fail. If the message posting fails, then this will skip the `terminate_instances` steps (since the post will throw an exception). However, if the initial post is moved after the `terminate_inst` step and the `terminate_inst` step fails then the post will never get sent. In either case, the "cull instances" cron job should catch both.

File: .github/workflows/firesim-cleanup.yml
Comment (abejgonzalez) [inline .github/workflows/firesim-cleanup.yml:5]: Just run this as much as possible.

File: .github/workflows/firesim-cleanup.yml
Comment (abejgonzalez) [inline .github/workflows/firesim-cleanup.yml:40]: Since this cron job isn't tied to a particular PR, just email the Slack channel directly.

File: .github/scripts/workflow-monitor.py
Comment (t14916) [reply .github/scripts/workflow-monitor.py:49]: I think it'd be fine just to comment on the issue in the case of a failure of the main loop. I think the checks you have in setup-workflow-monitor.py are sufficient to make sure it's up.

File: .github/scripts/setup-workflow-monitor.py
Comment (t14916) [inline .github/scripts/setup-workflow-monitor.py:30]:
```suggestion
        # run("echo 'zombie kr' >> ~/.screenrc") # for testing purposes, keep the screen on even after it dies
```

Comment (t14916) [reply .github/scripts/setup-workflow-monitor.py:30]: your check below will fail because with `zombie kr` enabled the screen session persists even on failure.

File: .github/workflows/firesim-cleanup.yml
Comment (t14916) [reply .github/workflows/firesim-cleanup.yml:40]: Which channel is this updating?

File: .github/scripts/setup-workflow-monitor.py
Comment (abejgonzalez) [reply .github/scripts/setup-workflow-monitor.py:30]: Ok, I will remove it. I was also checking if the screen invocation itself was incorrect (in addition to if the underlying session was still open).

**Comment by abejgonzalez** [comment]
Workflow monitor started for CI run: 3634757175

File: .github/scripts/workflow-monitor.py
Comment (abejgonzalez) [reply .github/scripts/workflow-monitor.py]: Done

File: .github/scripts/setup-workflow-monitor.py
Comment (abejgonzalez) [reply .github/scripts/setup-workflow-monitor.py]: Done

File: .github/scripts/workflow-monitor.py
Comment (abejgonzalez) [reply .github/scripts/workflow-monitor.py]: Done

File: .github/scripts/ci_variables.py
Comment (davidbiancolin) [reply .github/scripts/ci_variables.py:40]: Like it would'be have been better to create a global `dataclass` that you could reference instead of using stringly typed maps...

File: .github/scripts/platform_lib.py
Comment (davidbiancolin) [inline .github/scripts/platform_lib.py:431]: @t14916  These should just throw a not implemented exception (essentially `super`'s implementation, to ensure you get around to implementing them)

Comment (t14916) [reply .github/scripts/platform_lib.py:431]: Yeah I didn't notice these, this should just return the super's not implemented error. I would remove both of these from the azure platform lib.

File: .github/scripts/ci_variables.py
Comment (t14916) [reply .github/scripts/ci_variables.py:40]: I don't think mypy can do that

Comment (abejgonzalez) [reply .github/scripts/ci_variables.py:40]: I'm rolling out an implementation of https://peps.python.org/pep-0589/ to check that the keys exist. I think that is sufficient?

File: .github/scripts/platform_lib.py
Comment (abejgonzalez) [reply .github/scripts/platform_lib.py:431]: Done

**Comment by abejgonzalez** [comment]
Future TODO:

- Verify that FPGA instances are tagged appropriately so that we can shut down instances with all the extra checks.

File: scripts/run-py-typecheck.sh
Comment (abejgonzalez) [inline scripts/run-py-typecheck.sh:14]: Maybe it would be good to move the CI type checking to a different script / CI job that happens before spawning the manager instance so that any typing issues (i.e. keys not found in the `ci_env` dict) must be addressed before CI is run?

Comment (abejgonzalez) [reply scripts/run-py-typecheck.sh:14]: I didn't do this because it 1. requires typing to be implemented for all CI scripts (very invasive) and 2. wanted feedback before executing on this.

Comment (t14916) [reply scripts/run-py-typecheck.sh:14]: Most of the CI code should be typed right? I remember adding typing to a lot of stuff (not everything), regardless I think this is fine for now but perhaps leave that to a future PR. Personally, I have type checked all the CI code at some point, it seems to be fine mostly barring some issues I was having with libraries.

File: .github/scripts/platform_lib.py
Comment (abejgonzalez) [inline .github/scripts/platform_lib.py:191]: Since we put the `GITHUB_RUN_ID` into the cluster name, we can then use this tag to see what FPGA instances are running.

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3652751564. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Exception Message:**
```
Parameter validation failed:
Invalid type for parameter Filters[0].Values, value: *3652751564*, type: <class 'str'>, valid types: <class 'list'>, <class 'tuple'>
```
**Traceback Message:**
```
Traceback (most recent call last):
  File "/home/centos/firesim/.github/scripts/./workflow-monitor.py", line 62, in main
    platform_lib.check_and_terminate_run_farm_instances(45, ci_env['GITHUB_RUN_ID'])
  File "/home/centos/firesim/.github/scripts/platform_lib.py", line 275, in check_and_terminate_run_farm_instances
    instances = self.find_run_farm_ci_instances(workflow_tag)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/centos/firesim/.github/scripts/platform_lib.py", line 194, in find_run_farm_ci_instances
    ci_instances = get_instances_with_filter(instances_filter, allowed_states=['*'])
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/centos/actions-runner-0/_work/firesim/firesim/deploy/awstools/awstools.py", line 456, in get_instances_with_filter
    instance_res = depaginated_boto_query(ec2_client, 'describe_instances', operation_params, 'Reservations')
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/centos/actions-runner-0/_work/firesim/firesim/deploy/awstools/awstools.py", line 61, in depaginated_boto_query
    for page in page_iterator:
  File "/opt/conda/envs/firesim/lib/python3.11/site-packages/botocore/paginate.py", line 255, in __iter__
    response = self._make_request(current_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/firesim/lib/python3.11/site-packages/botocore/paginate.py", line 334, in _make_request
    return self._method(**current_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/firesim/lib/python3.11/site-packages/botocore/client.py", line 391, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/firesim/lib/python3.11/site-packages/botocore/client.py", line 691, in _make_api_call
    request_dict = self._convert_to_request_dict(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/firesim/lib/python3.11/site-packages/botocore/client.py", line 739, in _convert_to_request_dict
    request_dict = self._serializer.serialize_to_request(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/firesim/lib/python3.11/site-packages/botocore/validate.py", line 360, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter Filters[0].Values, value: *3652751564*, type: <class 'str'>, valid types: <class 'list'>, <class 'tuple'>

```

File: .github/scripts/ci_variables.py
Comment (davidbiancolin) [reply .github/scripts/ci_variables.py]: Sure, i'm curious why you chose that, but either implementation is much better.

File: .github/scripts/launch-azure-vm.py
Comment (davidbiancolin) [inline .github/scripts/launch-azure-vm.py:39]:
```suggestion
    # Networking related variables
```

File: scripts/run-py-typecheck.sh
Comment (davidbiancolin) [reply scripts/run-py-typecheck.sh:14]: I would really like to see mypy run against the CI scripts. If you want to divvy up the work i can help out a bit.

**Review by davidbiancolin — COMMENTED** [review]
My concerns have been addressed.

**Comment by davidbiancolin** [comment]
Deferring to @t14916 for final approval.

**Review by davidbiancolin — APPROVED** [review]

**Comment by t14916** [comment]
I think there is still an issue with the workflow monitor, so I guess I'll withhold approval until that is fixed (but everything else lgtm)

**Comment by abejgonzalez** [comment]
> I think there is still an issue with the workflow monitor, so I guess I'll withhold approval until that is fixed (but everything else lgtm)

Agreed. I missed a `[]`. I'll make sure we get full checkmarks before merge.

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3654677919. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Exception Message:**
```
'dict' object has no attribute 'launch_time'
```
**Traceback Message:**
```
Traceback (most recent call last):
  File "/home/centos/firesim/.github/scripts/./workflow-monitor.py", line 62, in main
    platform_lib.check_and_terminate_run_farm_instances(45, ci_env['GITHUB_RUN_ID'])
  File "/home/centos/firesim/.github/scripts/platform_lib.py", line 278, in check_and_terminate_run_farm_instances
    if (datetime.datetime.now() - inst.launch_time) >= datetime.timedelta(minutes=timeout):
AttributeError: 'dict' object has no attribute 'launch_time'

```

**Comment by abejgonzalez** [comment]
Something went wrong in the workflow monitor for CI run 3655073569. Verify CI instances are terminated properly. Must be checked before submitting the PR.

**Exception Message:**
```
can't subtract offset-naive and offset-aware datetimes
```
**Traceback Message:**
```
Traceback (most recent call last):
  File "/home/centos/firesim/.github/scripts/./workflow-monitor.py", line 62, in main
    platform_lib.check_and_terminate_run_farm_instances(45, ci_env['GITHUB_RUN_ID'])
  File "/home/centos/firesim/.github/scripts/platform_lib.py", line 278, in check_and_terminate_run_farm_instances
    if (datetime.datetime.now() - inst['LaunchTime']) >= datetime.timedelta(minutes=timeout):
        ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
TypeError: can't subtract offset-naive and offset-aware datetimes

```

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
Uncaught 2 FPGA instance shutdown(s) detected for CI run: 3659621259. Verify CI state before submitting PR.

**Comment by abejgonzalez** [comment]
@t14916 This should fully work now. Sorry for the delay.

**Review by t14916 — APPROVED** [review]
LGTM

**Pull request merged**
